### PR TITLE
[PyROOT] Migrate RDataFrame, RVec and RTensor pythonizations to Python

### PR DESCRIPTION
**Pull request by guitargeek**
Like this, we are using fewer CPyCppyy implementation details in the pythonizations.

### CODE REVIEW

**Comment by guitargeek** [comment]
Ah! My bad, I didn't notice that the code was still used!

My `git grepping` didn't give me a git because the names of the C++ function and the name of the function on the Python side is different :slightly_smiling_face: 
https://github.com/root-project/root/blob/master/bindings/pyroot/pythonizations/src/PyROOTModule.cxx#L73
```c++
   {(char *)"MakeNumpyDataFrame", (PyCFunction)PyROOT::MakeNumpyDataFrameImpl, METH_O,
    (char *)"Make RDataFrame from dictionary of numpy arrays"},
```

Anyway, I'll keep this PR open and maybe see later how this function can be re-implemented in pure Python.

File: bindings/pyroot/pythonizations/python/ROOT/_facade.py [hunk 1]
```diff
@@ -9,13 +9,186 @@
...
+            """
+std::string PyROOT_getByteswap()
+{
+#ifdef R__BYTESWAP
+    return "<";
+#else
+    return ">";
+#endif
+}
+"""
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_facade.py:39]: This is not nice to me. I would much rather we had a function somewhere in ROOT (e.g. core/foundation) that we can call to get this information without asking the interpreter

File: bindings/pyroot/pythonizations/python/ROOT/_facade.py [hunk 2]
```diff
@@ -9,13 +9,186 @@
...
+        raise RuntimeError('Failed to find "ROOT/RNumpyDS.hxx".')
+
+    address = ROOT.gInterpreter.Calc(code)
+    rdf = cppyy.bind_object(address, "ROOT::RDataFrame")
+    ROOT.SetOwnership(rdf, True)
+
+    # Bind pyobject holding adopted memory to the RVec
+    setattr(rdf, "__data__", pyvecs)
+
+    return rdf
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_facade.py:190]: All of this code does not belong to `_facade.py`, should be moved to some helper module

File: bindings/pyroot/pythonizations/python/ROOT/_facade.py [hunk 3]
```diff
@@ -9,13 +9,186 @@
...
 
 from ._pythonization import pythonization
 
+byteswap = ""
+
+
+def _CheckEndianessFromTypestr(typestr):
+    import ROOT
+
+    global byteswap
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_facade.py:26]: No `global` please

File: bindings/pyroot/pythonizations/python/ROOT/_facade.py [hunk 4]
```diff
@@ -9,13 +9,186 @@
...
+    elif dtype == "i8":
+        return "Long64_t"
+    elif dtype == "u8":
+        return "ULong64_t"
+    elif dtype == "f4":
+        return "float"
+    elif dtype == "f8":
+        return "double"
+
+    raise RuntimeError("Object not convertible: Python object has unknown data-type '" + dtype + "'.")
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_facade.py:65]: This is a good chance to convert this function to a Python dictionary which would be much more idiomatic

File: bindings/pyroot/pythonizations/python/ROOT/_facade.py [hunk 5]
```diff
@@ -9,13 +9,186 @@
...
+    import ROOT
+
+    if not isinstance(np_dict, dict):
+        raise RuntimeError("Object not convertible: Python object is not a dictionary.")
+
+    if len(np_dict) == 0:
+        raise RuntimeError("Object not convertible: Dictionary is empty.")
+
+    is_windows = os.name == "nt"
+    address_prefix = "0x" if is_windows else ""
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_facade.py:147]:
```suggestion
    address_prefix = "0x" if platform.system() == "Windows" else ""
```

https://docs.python.org/3/library/platform.html#platform.system

**Review by vepadulano — CHANGES_REQUESTED** [review]
Thanks! I like the idea, I left some initial comments about things I believe should be improved

File: bindings/pyroot/pythonizations/python/ROOT/_facade.py (hunk 3 above)
Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_facade.py]: I got rid of it and made it a data member of the facase (for now :))

File: bindings/pyroot/pythonizations/python/ROOT/_facade.py (hunk 2 above)
Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_facade.py]: That's true. I moved it to `_rvec`, `_rtensor`, and `_rdataframe`.

File: bindings/pyroot/pythonizations/python/ROOT/_facade.py (hunk 1 above)
Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_facade.py]: You are right, it's not optimal. However, I would like to keep this PR restricted to the PyROOT code area.

Therefore, I have put back the `GetEndianess()` function in the C++ `libROOTPythonizations` for now, and then we can think later how to "upstream" this to ROOT.

**Comment by guitargeek** [comment]
Hi @vepadulano, thanks a lot for reviewing this PR! I have addressed your comments, and also migrated the `AsRTensor` function to Python, in the same way as the `AsRVec` function.

The changes after the review comments are for now appended as fixup commits, but at the end of the review I will squash the commits.

**Comment by hahnjo** [comment]
@phsft-bot build!

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py [hunk 6]
```diff
@@ -96,17 +165,15 @@ def get_array_interface(self):
             else:
                 pointer = GetDataPointer(self, cppname, "data")
             return {
-                "shape": (size, ),
+                "shape": (size,),
                 "typestr": "{}{}{}".format(endianess, dtype_numpy, dtype_size),
                 "version": 3,
-                "data": (pointer, False)
+                "data": (pointer, False),
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py:171]: These cosmetic changes could go a in a separate commit

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py [hunk 7]
```diff
@@ -96,17 +165,15 @@ def get_array_interface(self):
...
             }
 
 
 def add_array_interface_property(klass, name):
-    if True in [
-            name.endswith("<{}>".format(dtype)) for dtype in _array_interface_dtype_map
-    ]:
+    if True in [name.endswith("<{}>".format(dtype)) for dtype in _array_interface_dtype_map]:
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py:176]: Cosmetic

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py [hunk 8]
```diff
@@ -423,3 +423,70 @@ def pythonize_rdataframe(klass):
...
+    \param[in] self Always null, since this is a module function.
+    \param[in] pydata Dictionary with numpy arrays
+
+    This function takes a dictionary of numpy arrays and creates an RDataFrame
+    using the keys as column names and the numpy arrays as data.
+    """
+    import ROOT
+    import platform
+
+    cppyy = ROOT.module.cppyy
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:441]: Why can't we `import cppyy` directly in this case?

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py [hunk 9]
```diff
@@ -9,10 +9,84 @@
...
+    prefix = "0x" if platform.system() == "Windows" else ""
+    klassname = f"TMVA::Experimental::RTensor<{cppdtype},std::vector<{cppdtype}>>"
+    code = f"new {klassname}(reinterpret_cast<{cppdtype} *>({prefix}{data}),"
+    code += "{" + ",".join([f"{s}" for s in shape]) + "}, "
+    code += "{" + ",".join([f"{s // dtypesize}" for s in strides]) + "}, "
+    if row_major:
+       code += "TMVA::Experimental::MemoryLayout::RowMajor"
+    else:
+       code += "TMVA::Experimental::MemoryLayout::ColumnMajor"
+    code += ")"
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py:78]: Bit of a nitpick, but I believe a multiline f-string would be more idiomatic here

**Review by vepadulano — COMMENTED** [review]
Thanks a lot! Some minor comments for consideration, I believe this could be already good to merge (I'll check the CI results tomorrow)

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py (hunk 9 above)
Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py:78]: Sure, why not!

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py (hunk 8 above)
Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:441]: That's true, I think it can be done

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py (hunk 6 above)
Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py:171]: I have put the cosmetic changes in a separate commit

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py (hunk 7 above)
Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py:176]: I have put the cosmetic changes in a separate commit

**Comment by guitargeek** [comment]
Hi @vepadulano, thanks for your follow-up review! I have implemented your comments. Let's hope the CI is green tomorrow morning :+1:

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py [hunk 10]
```diff
@@ -9,10 +9,83 @@
...
+    klassname = f"TMVA::Experimental::RTensor<{cppdtype},std::vector<{cppdtype}>>"
+    address = ROOT.gInterpreter.Calc(
+        f"new {klassname}(reinterpret_cast<{cppdtype} *>({prefix}{data}),"
+        "{" + ",".join([f"{s}" for s in shape]) + "}, "
+        "{" + ",".join([f"{s // dtypesize}" for s in strides]) + "}, "
+        "TMVA::Experimental::MemoryLayout::" + memory_layout + ")"
+    )
+
+    # Bind the object to a Python-side proxy, taking ownership of the underlying C++ object
+    out = cppyy.bind_object(address, klassname)
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py:80]: Can we make these lines even more Pythonic? Something like the following (untested pseudocode)
```suggestion
    tensor = TMVA.RTensor[cppdtype, std.vector[cppdtype]](
        cppyy.ll.reinterpret_cast[f"{cppdtype} *"](f"{prefix}{data}"),
        shapetuple,
        memory_layout
    )
```

Then I guess also the ownership will be automatically handled if this works

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py [hunk 11]
```diff
@@ -82,6 +82,75 @@
...
+    # Construct an RVec of the correct data-type
+    klassname = "ROOT::VecOps::RVec<" + cppdtype + ">"
+
+    prefix = "0x" if platform.system() == "Windows" else ""
+
+    address = ROOT.gInterpreter.Calc(f"new {klassname}(reinterpret_cast<{cppdtype} *>({prefix}{data}), {size})")
+
+    # Bind the object to a Python-side proxy, taking ownership of the underlying C++ object
+    out = cppyy.bind_object(address, klassname)
+    ROOT.SetOwnership(out, True)
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py:146]: Same comment as with `_rtensor`, it would be very nice if we can avoid the extra call to the interpreter

**Review by vepadulano — CHANGES_REQUESTED** [review]

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py (hunk 10 above)
Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py]: Thanks a lot, very good suggestions! I implemented them. Let's see what the CI says now.

Unfortunately, I still have a Windows failure that I still don't understand. Maybe it will be magically gone after implementing your suggestions :)

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py [hunk 12]
```diff
@@ -82,6 +82,61 @@
...
+            "Object not convertible: __array_interface__['typestr'] returned '"
+            + typestr
+            + "' with invalid length unequal 3."
+        )
+
+    dtype = typestr[1:]
+    cppdtype = _get_cpp_type_from_numpy_type(dtype)
+
+    # Construct an RVec of the correct data-type
+    out = ROOT.VecOps.RVec[cppdtype](ROOT.module.cppyy.ll.reinterpret_cast[f"{cppdtype} *"](data), [size])
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py:132]: Shouldn't this be?
```suggestion
    out = ROOT.VecOps.RVec[cppdtype](ROOT.module.cppyy.ll.reinterpret_cast[f"{cppdtype} *"](data), size)
```

Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py:132]: Right, thanks

**Comment by guitargeek** [comment]
Sorry I had to update it so many times! ROOT `master` changes so much in PyROOT, I needed to update more than just fixing git conflicts...

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py [hunk 13]
```diff
@@ -402,24 +403,89 @@ def pythonize_rdataframe(klass):
...
+
+        return code
+
+    # Iterate over dictionary, convert numpy arrays to RVecs and put together interpreter code
+    code += ", ".join([write_vec_code(key, value) for key, value in np_dict.items()]) + ")"
+
+    # Create RDataFrame and build Python proxy
+    err = ROOT.gInterpreter.Declare('#include "ROOT/RNumpyDS.hxx"')
+    if not err:
+        raise RuntimeError('Failed to find "ROOT/RNumpyDS.hxx".')
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:482]: Is this even needed? In a ROOT build with RDataFrame and PyROOT enabled, `RNumpyDS.hxx` should already be there? We could remove another call to the interpreter

Comment (vepadulano) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:482]: Clearly this is much less urgent than other places since it's done only once at construction of the RDataFrame object

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py [hunk 14]
```diff
@@ -402,24 +403,89 @@ def pythonize_rdataframe(klass):
...
+
+    address = ROOT.gInterpreter.Calc(code)
+    rdf = cppyy.bind_object(address, "ROOT::RDataFrame")
+    ROOT.SetOwnership(rdf, True)
+
+    # Bind pyobject holding adopted memory to the RVec
+    setattr(rdf, "__data__", pyvecs)
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:489]:
```suggestion
    # Bind pyobject holding adopted memory to the RVec
   rdf.__data__ = pyvecs
```

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py [hunk 15]
```diff
@@ -82,6 +82,61 @@
...
+        )
+
+    dtype = typestr[1:]
+    cppdtype = _get_cpp_type_from_numpy_type(dtype)
+
+    # Construct an RVec of the correct data-type
+    out = ROOT.VecOps.RVec[cppdtype](ROOT.module.cppyy.ll.reinterpret_cast[f"{cppdtype} *"](data), size)
+
+    # Bind pyobject holding adopted memory to the RVec
+    setattr(out, "__adopted__", arr)
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py:135]:
```suggestion
    out.__adopted__ = arr
```

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py [hunk 16]
```diff
@@ -7,12 +7,72 @@
...
+    dtypesize = int(typestr[-1])
+    cppdtype = _get_cpp_type_from_numpy_type(dtype)
+
+    # Get strides
+    strides = arr.strides
+
+    # Infer memory layout from strides
+    memory_layout = ROOT.TMVA.Experimental.MemoryLayout.RowMajor
+    if len(strides) > 1 and strides[0] < strides[-1]:
+        memory_layout = ROOT.TMVA.Experimental.MemoryLayout.ColumnMajor
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py:60]: Make this a ternary operator assignment maybe

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py [hunk 17]
```diff
@@ -7,12 +7,72 @@
...
+    # Construct an RTensor of the correct data-type
+    out = ROOT.TMVA.Experimental.RTensor[cppdtype, ROOT.std.vector[cppdtype]](
+        ROOT.module.cppyy.ll.reinterpret_cast[f"{cppdtype} *"](data),
+        [s for s in shape],
+        [s // dtypesize for s in strides],
+        memory_layout,
+    )
+
+    # Bind pyobject holding adopted memory to the RTensor
+    setattr(out, "__adopted__", arr)
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py:71]:
```suggestion
    out.__adopted__ = arr
```

**Review by vepadulano — CHANGES_REQUESTED** [review]
Super good! A few final comments for consideration before final approval

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py (hunk 13 above)
Comment (guitargeek) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:482]: Interesting observation, but for now it doesn't work without removing this line. I think this is something for a follow-up PR.

**Comment by guitargeek** [comment]
Hi @vepadulano! I have addressed your latest comments, thanks for the review :+1:

**Review by vepadulano — APPROVED** [review]
Amazing!!! Thanks a lot, let's merge this :rocket:
